src/lxc/commands.c

```diff
@@ -932,6 +955,105 @@ static int lxc_cmd_add_state_client_callback(int fd, struct lxc_cmd_req *req,
 	return lxc_cmd_rsp_send(fd, &rsp);
 }
 
+int lxc_cmd_openpty(const char *name, const char *lxcpath, int *amaster,
+		    int *aslave)
+{
+	int ret, stopped;
+	struct lxc_cmd_rr cmd = {
+	    .req = {
+		    .cmd = LXC_CMD_OPENPTY
+	    },
+	};
+	struct lxc_cmd_openpty_rsp_data *openpty_rsp;
+
+	ret = lxc_cmd(name, &cmd, &stopped, lxcpath, NULL);
+	if (ret < 0)
+		return -1;
+
+	if (cmd.rsp.ret < 0)
+		return -1;
+
+	openpty_rsp = cmd.rsp.data;
+
+	*amaster = openpty_rsp->master;
+	*aslave = openpty_rsp->slave;
+
+	return 0;
+}
+
+static int lxc_cmd_openpty_callback(int fd, struct lxc_cmd_req *req,
+				    struct lxc_handler *handler)
+{
+#ifdef TIOCGPTPEER
+	int ret;
+	struct lxc_cmd_rsp rsp;
+	char proc_self_path[LXC_PROC_SELF_LEN];
+	int pty_fds[2] = {-1, -1};
+
+	memset(&rsp, 0, sizeof(rsp));
+
+	ret = snprintf(proc_self_path, LXC_PROC_SELF_LEN, "/proc/self/fd/%d",
+		       handler->conf->dev_ptmx.opath_fd);
+	if (ret < 0 || (size_t)ret >= LXC_PROC_SELF_LEN) {
+		SYSERROR("Failed to create string: %d != %d", ret,
+			 LXC_PROC_SELF_LEN);
+		return -1;
+	}
+
+	/* open the pseudoterminal master */
+	pty_fds[0] = open(proc_self_path, O_RDWR | O_NOCTTY);
+	if (pty_fds[0] < 0) {
+		ERROR("%s - Failed to open \"%s\"", proc_self_path,
+		      strerror(errno));
+		return -1;
+	}
+
+	ret = grantpt(pty_fds[0]);
```

Comment:
This needs to be replaced by our own `grantpt()` implementation since libcs don't guarantee that this is not a path-based operation. Instead we should operate directly on the fd itself which shouldn't be too difficult to do.